File: 27_remove_element/remove_element.cpp
/*
 * @Date: 2021-04-19 08:38:32
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-04-19 08:52:30
 */

#include <iostream>
#include <vector>
using namespace std;

void print_vec(const vector<int>& nums, const int& size) {
  for (int i = 0; i < size; i++) cout << nums[i] << ", ";
  cout << endl;
}

int remove_element(vector<int>& nums, int val) {
  if (nums.size() == 0) return 0;
  int j = 0;
  for (int i = 0; i < nums.size(); i++) {
    if (nums[i] != val) {
      nums[j++] = nums[i];
    }
  }
  return j;
}

int main() {
  {
    vector<int> nums{3, 2, 2, 3};
    int size = remove_element(nums, 3);
    print_vec(nums, size);
  }
  {
    vector<int> nums{0, 1, 2, 2, 3, 0, 4, 2};
    int size = remove_element(nums, 2);
    print_vec(nums, size);
  }
}

File: 27_remove_element/remove_element.rs
/*
 * @Date: 2021-04-19 08:38:23
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-04-19 09:11:06
 */

fn remove_element(nums: &mut Vec<i32>, val: i32) -> i32 {
    if nums.len() == 0 {
        return 0;
    }
    let mut j = 0;
    for i in 0..nums.len() {
        if nums[i] != val {
            nums[j] = nums[i];
            j += 1;
        }
    }
    j as i32
}

fn main() {
    {
        let mut nums = vec![3, 2, 2, 3];
        assert_eq!(remove_element(&mut nums, 3), 2);
        println!("{:?}", nums);
    }
    {
        let mut nums = vec![0, 1, 2, 2, 3, 0, 4, 2];
        assert_eq!(remove_element(&mut nums, 2), 5);
        println!("{:?}", nums);
    }
}
